docs/hd.md

```diff
@@ -1,124 +1,212 @@
+wallet tree because decrypted derivation paths describe it.
+
+So now it's obviously that if you share your `RootPK` - you give to anyone an
+ability to reveal all your addresses in the blockchain!
 
 ## Transaction creation
 
-To spend `TxIn` wallet application must sign them using secret key and provide
-public key as a witness.
-For HD wallets we do the same thing: we derive secret and public key of
-address corresponding to `TxIn` from root secret key using user's passphrase
-and sign `TxIn` by derived secret key and put public key as a witness.
-
-Note: although we can reveal all addresses just knowing root public key,
-to spend `TxIn`s root secret key and user's passphrase are needed.
+Suppose that we derived `dSK_A` and generated a new address `A` from it. Some
+other user `OU` sent 100 ADA to the address `A`. Technically it means that `OU`
+created a new transaction `T1` and its output `O` contains an address `A` and
+this amount.
+
+Now we want to send these 100 ADA to another user `AU`. Technically it means
+that we have to create a new transaction `T2` and its input `I` will refer to
+an output `O` of `T1`:
+
+        T1               T2
+    +--------+       +--------+
+    |        |       |        |
+    |    +---+       +---+    |
+    |    | O |------>| I |    |
+    |    +---+       +---+    |
+    |        |       |        |
+    +--------+       +--------+
+
+Now we must prove that we have a right to spend money from the address `A`
+(which is in `O`). To do it we must prove that address `A` was generated by
+our wallet (because if so, money on this address is ours and we obviously
+may spend it).
+
+This is where derivation path helps us: since address `A` contains this path,
+we can take it and derive `dSK_A` and corresponding `dPK`. This is how we prove
+that address `A` is ours: only we could derive `dPK`, which means that only
+we could generate address `A`.
+
+Now we have to create a _witness_ for each input of transaction. A witness is
+our proof that we may spend money from corresponding address, it includes
+a signature `S` and public key `PK`:
+
+1. `S` is a signature of entire transaction (technically it's a signature of
+   the hash of the transaction),
+2. `PK` is a derived public key that corresponds to derived secret key which
+   was used to generate our address.
+
+In our example address `A` was generated from `dSK_A`, so witness for `I` will
+include:
+
+1. A signature of the transaction `T2`,
+2. Public key `dPK_A` that corresponds to `dSK_A`.
+
+Please note: although we can reveal all wallet's addresses just knowing `RootPK`
+of this wallet, to spend money `RootSK` is needed.
+
+## Formal specification
+
+Please note that the purpose of this document is to provide a high-level overview
+of HD wallets. For technical details please read [Formal specification for a Cardano
+wallet](https://cardanodocs.com/files/formal-specification-of-the-cardano-wallet.pdf).
```

Comment:
I have no idea why this commit shows up in my PR. Perhaps GitHub is confused? I did rebase develop recently, so maybe the commit got picked up...